File: LeetCode/Java Solutions/Intersection of Two Arrays/intersection.java
class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        // Use HashSet for removing duplicates and for efficient lookups
        Set<Integer> set1 = new HashSet<>();
        for (int num : nums1) {
            set1.add(num);
        }
        
        Set<Integer> intersection = new HashSet<>();
        for (int num : nums2) {
            if (set1.contains(num)) {
                intersection.add(num);
            }
        }
        
        // Convert the intersection set to an array
        int[] result = new int[intersection.size()];
        int i = 0;
        for (Integer num : intersection) {
            result[i++] = num;
        }
        
        return result;
    }
}

File: LeetCode/Java Solutions/Intersection of Two Arrays II/intersection.py
from typing import List
from collections import Counter

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Count the occurrences of each element in both arrays
        count1 = Counter(nums1)
        count2 = Counter(nums2)
        
        # Find the intersection of both counters
        intersection = count1 & count2
        
        # Convert the intersection to a list of elements
        result = []
        for num in intersection.elements():
            result.append(num)
        
        return result

# Example usage:
sol = Solution()
print(sol.intersect([1, 2, 2, 1], [2, 2]))  # Output: [2, 2]
print(sol.intersect([4, 9, 5], [9, 4, 9, 8, 4]))  # Output: [4, 9]
